sys/stdio_nimble/stdio_nimble.c

```diff
@@ -0,0 +1,391 @@
+            _debug_printf("BLE_GAP_EVENT_DISCONNECT\n");
+            _status = STDIO_NIMBLE_DISCONNECTED;
+            break;
+
+        case BLE_GAP_EVENT_SUBSCRIBE:
+            _debug_printf("BLE_GAP_EVENT_SUBSCRIBE\n");
+            if (event->subscribe.attr_handle == _val_handle_stdout) {
+                if (event->subscribe.cur_indicate == 1) {
+                    _status = STDIO_NIMBLE_SUBSCRIBED;
+                }
+                else {
+                    _status = STDIO_NIMBLE_CONNECTED;
+                }
+            }
+            break;
+
+        case BLE_GAP_EVENT_NOTIFY_TX:
+            _debug_printf("BLE_GAP_EVENT_NOTIFY_TX\n");
+            if (event->notify_tx.indication == 1) {
+                if (event->notify_tx.status == BLE_HS_EDONE) {
+                    _status = STDIO_NIMBLE_SUBSCRIBED;
+                }
+                else if (event->notify_tx.status != 0) {
+                    _status = STDIO_NIMBLE_SUBSCRIBED;
+                }
+            }
+            break;
+
+        case BLE_GAP_EVENT_MTU:
+            _debug_printf("BLE_GAP_EVENT_MTU: mtu = %d\n", event->mtu.value);
+            break;
+    }
+
+    return 0;
+}
+
+static int gatt_svr_chr_access_stdin(
+        uint16_t conn_handle, uint16_t attr_handle,
+        struct ble_gatt_access_ctxt *ctxt, void *arg)
+{
+    (void) conn_handle;
+    (void) attr_handle;
+    (void) arg;
+
+    uint16_t om_len = OS_MBUF_PKTLEN(ctxt->om);
+
+    /* read sent data */
+    int rc = ble_hs_mbuf_to_flat(ctxt->om, _stdin_read_buf,
+            sizeof(_stdin_read_buf), &om_len);
+
+    isrpipe_write(&_isrpipe_stdin, _stdin_read_buf, om_len);
+
+    return rc;
+}
+
+void stdio_init(void)
+{
+#if IS_USED(MODULE_VFS)
+    vfs_bind_stdio();
+#endif
+
+#if IS_USED(MODULE_STDIO_NIMBLE_DEBUG)
+    uart_init(STDIO_UART_DEV, STDIO_UART_BAUDRATE, NULL, NULL);
+#endif
+
+    ble_npl_callout_init(&_send_stdout_callout, nimble_port_get_dflt_eventq(),
+                         _send_stdout, NULL);
+}
+
+ssize_t stdio_read(void *buffer, size_t count)
+{
+    /* blocks until at least one character was read */
+    ssize_t res = isrpipe_read(&_isrpipe_stdin, buffer, count);
+
+#if IS_USED(MODULE_STDIO_NIMBLE_DEBUG)
+    mutex_lock(&_debug_print_mutex);
+    uart_write(STDIO_UART_DEV, (const uint8_t *)PREFIX_STDIN, strlen(PREFIX_STDIN));
+    uart_write(STDIO_UART_DEV, (const uint8_t *)buffer, res);
+    uart_write(STDIO_UART_DEV, (const uint8_t *)"\n", 1);
+    mutex_unlock(&_debug_print_mutex);
+#endif
+
+    return res;
+}
+
+ssize_t stdio_write(const void* buffer, size_t len)
+{
+#if IS_USED(MODULE_STDIO_NIMBLE_DEBUG)
+    mutex_lock(&_debug_print_mutex);
+    uart_write(STDIO_UART_DEV, (const uint8_t *)PREFIX_STDOUT, strlen(PREFIX_STDOUT));
+    uart_write(STDIO_UART_DEV, (const uint8_t *)buffer, len);
+    uart_write(STDIO_UART_DEV, (const uint8_t *)"\n", 1);
+    mutex_unlock(&_debug_print_mutex);
+#endif
+
+    mutex_lock(&_tsrb_stdout_mutex);
```

Comment:
Why  a mutex? Can't we simply enable/disable irq?

Reply:
No strong objections against it, personally I simply consider enabling and disabling of irq quite low-level even within embedded development. Are we sure disabling irq for prints has no side effects on the nimble thread?

Reply:
Well `irq_disable` can always have an impact, and it should be for short periods of time, but `tsrb_add` is already doing so, so the impact wouldn't be much of doing this instead of `mutex_lock/unlock`. If this is a concern, then maybe that was the reason you avoided `tsrb` in the initial implementation? If thats the case it would be an issue with `tsrb` when wanting to use `tsrb_add` on big chunks.

Reply:
The other thing is that can't `stdio_write` be called from an ISR? In that case you could end up in a deadlock here, hich would be a reson not to use `mutex`